internal/credit/creditdriver/balance.go

```diff
@@ -61,3 +66,73 @@ func (b *builder) GetLedgerBalance() GetLedgerBalanceHandler {
 			}))...,
 	)
 }
+
+func mapBalanceToAPI(balance credit.Balance, highwatermark credit.HighWatermark) api.LedgerBalance {
+	var featureBalances []api.FeatureBalance = make([]api.FeatureBalance, 0, len(balance.FeatureBalances))
+	for _, featureBalance := range balance.FeatureBalances {
+		featureBalances = append(featureBalances, mapFeatureBalanceToAPI(featureBalance))
+	}
+
+	var grantBalances []api.LedgerGrantBalance = make([]api.LedgerGrantBalance, 0, len(balance.GrantBalances))
+	for _, grantBalance := range balance.GrantBalances {
+		grantBalances = append(grantBalances, mapGrantBalanceToAPI(grantBalance))
+	}
+
+	return api.LedgerBalance{
+		FeatureBalances: featureBalances,
+		GrantBalances:   grantBalances,
+		LastReset:       &highwatermark.Time,
+		Metadata:        &balance.Metadata,
+		Subject:         balance.Subject,
+	}
+}
+
+func mapFeatureBalanceToAPI(featureBalance credit.FeatureBalance) api.FeatureBalance {
+	id := new(string)
+	if featureBalance.ID != nil {
+		*id = string(*featureBalance.ID)
+	}
+	return api.FeatureBalance{
+		Archived:            featureBalance.Archived,
+		CreatedAt:           featureBalance.CreatedAt,
+		Id:                  id,
+		MeterGroupByFilters: featureBalance.MeterGroupByFilters,
+		MeterSlug:           featureBalance.MeterSlug,
+		Name:                featureBalance.Name,
+		UpdatedAt:           featureBalance.UpdatedAt,
+		Usage:               featureBalance.Usage,
+		Balance:             featureBalance.Balance,
+	}
+}
+
+func mapGrantBalanceToAPI(grantBalance credit.GrantBalance) api.LedgerGrantBalance {
+	var id *string
```

Comment:
For these let's have a generic like this (and put that into a package please, e.g convert):

```
package main

import "fmt"

type TestID string

func ConvertString[Source, Dest ~string](src *Source) *Dest {
	if src == nil {
		return nil
	}

	out := Dest(*src)
	return &out
}

func main() {

	ID := TestID("test")
	fmt.Println(*ConvertString[TestID, string](&ID))
}
```

An alternative could be that in the openapi spec at least for the ID types we are referencing the credit types. That would make our life slightly easier and would not allow for any divergence on the struct contents.

Reply:
i wasn't aware Dest ~string ... Dest(x) is possible in go 🤯